src/tests/ftest/launch.py

```diff
@@ -1933,6 +1939,74 @@ def finalize_bullseye(self):
                             os.rename(old_file, new_file)
         return status
 
+    def setup_slurm(self):
+        """Set up slurm on the hosts if any tests are using partitions.
+
+        Returns:
+            int: status code: 0 = success, 128 = failure
+        """
+        status = 0
+        logger.info("Setting up slurm partitions if required by tests")
+        if not any(test.yaml_info["client_partition"] for test in self.tests):
+            logger.debug("  No tests using client partitions detected - skipping slurm setup")
+            return status
+
+        if not self.slurm_add_partition:
+            logger.debug("  The 'slurm_setup' argument is not set - skipping slurm setup")
+            return status
+
+        status |= self.setup_application_directory()
+
+        slurm_setup = SlurmSetup(logger, self.slurm_partition_hosts, self.slurm_control_node, True)
+        try:
+            slurm_setup.install()
+            slurm_setup.update_config(self.user, 'daos_client')
+            slurm_setup.start_munge(self.user)
+            slurm_setup.start_slurm(self.user, True)
+        except SlurmSetupException:
+            message = "Error setting up slurm"
+            self._fail_test(self.result.tests[-1], "Run", message, sys.exc_info())
+            status |= 128
+
+        return status
+
+    def setup_application_directory(self):
+        """Set up the application directory.
+
+        Returns:
+            int: status code: 0 = success, 128 = failure
+        """
+        app_dir = os.environ.get('DAOS_TEST_APP_DIR')
+        app_src = os.environ.get('DAOS_TEST_APP_SRC')
+
+        logger.debug('Setting up the \'%s\' application directory', app_dir)
+        if not os.path.exists(app_dir):
+            # Create the apps directory if it does not already exist
+            try:
+                logger.debug('  Creating the application directory')
+                os.makedirs(app_dir)
+            except OSError:
+                message = 'Error creating the application directory'
+                self._fail_test(self.result.tests[-1], 'Run', message, sys.exc_info())
+                return 128
+        else:
+            logger.debug('  Using the existing application directory')
+
+        if app_src and os.path.exists(app_src):
+            logger.debug('  Copying applications from the \'%s\' directory', app_src)
+            run_local(logger, f'ls -al {app_src}')
+            for app in os.listdir(app_src):
+                try:
+                    run_local(logger, f'cp -r {os.path.join(app_src, app)} {app_dir}', check=True)
```

Comment:
Should we quote these paths?
```suggestion
                    run_local(logger, f'cp -r "{os.path.join(app_src, app)}" "{app_dir}"', check=True)
```

Reply:
Done